src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -96,12 +97,107 @@ public static void Fill(Span<byte> data)
+            if (adjustedRange == 1)
+            {
+                return fromInclusive;
+            }
+
+            // Subtract one to account for zero being an actual possibility. We know for certain that
+            // the range fits in a UInt32 without overflowing.
+            uint maskedRange = (uint)adjustedRange - 1;
+
+            int leadingZeroCount = LeadingZeroCount(maskedRange);
+
+            // Create a mask for the bits that we care about for the range. The other bits will be
+            // masked away.
+            uint mask;
+
+            // If the LZCNT is zero create a full mask to avoid overflowing when shifting.
+            if (leadingZeroCount == 0)
+            {
+                mask = unchecked((uint)-1);
+            }
+            else
+            {
+                mask = (1U << 32 - leadingZeroCount) - 1;
+            }
+
+            // We only want to generate as many bytes as required to satisfy the mask to not apply
+            // undue pressure to the underlying random number generator.
+            int bytesRequired = (32 + 7 - leadingZeroCount) / 8;
+
+            Span<uint> valueResult = stackalloc uint[1];
+            ref uint result = ref valueResult[0];
+            Span<byte> valueResultBytes = MemoryMarshal.AsBytes(valueResult);
+            if (BitConverter.IsLittleEndian)
+            {
+                do
+                {
+                    Span<byte> randomFillSlice = valueResultBytes.Slice(0, bytesRequired);
+                    RandomNumberGeneratorImplementation.FillSpan(randomFillSlice);
+                }
+                while ((result & mask) > maskedRange);
+            }
+            else
+            {
+                do
+                {
+                    Span<byte> randomFillSlice = valueResultBytes.Slice(bytesRequired);
+                    RandomNumberGeneratorImplementation.FillSpan(randomFillSlice);
+                }
+                while ((result & mask) > maskedRange);
+            }
+            return (int)((result & mask) + fromInclusive);
+        }
+
+        public static int GetInt32(int toExclusive)
+        {
+            if (toExclusive <= 0) throw new ArgumentOutOfRangeException(nameof(toExclusive), SR.ArgumentOutOfRange_NeedPosNum);
+            return GetInt32(0, toExclusive);
+        }
+
         internal void VerifyGetBytes(byte[] data, int offset, int count)
         {
             if (data == null) throw new ArgumentNullException(nameof(data));
             if (offset < 0) throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNum);
             if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
             if (count > data.Length - offset) throw new ArgumentException(SR.Argument_InvalidOffLen);
         }
+
+        private static int LeadingZeroCount(uint value)
+        {
+            //TODO: Use lzcnt intrinsics when available.
```

Comment:
I wasn't sure if it was acceptable at the current point to add a reference to `System.Runtime.Intrinsics.Experimental`.